packages/utils-algebra/lib/utils.ts

```diff
@@ -78,18 +86,179 @@ const transformer = new TransformerSubTyped<KnownOperation>({
   [Types.COPY]: { ignoreKeys: new Set([ 'source', 'destination', 'metadata' ]) },
 });
 
-export const mapOperation: (typeof transformer.transformNode<'unsafe', Operation>) = <any>
-  transformer.transformNode.bind(transformer);
-
+/**
+ * Transform a single operation, similar to {@link mapOperation}, but using stricter typings.
+ * e.g. wrapping a distinct around the outermost project:
+ * ```ts
+ * mapOperationStrict<'unsafe', Operation>(
+ *   { type: 'slice', input: { type: 'project', input: { type: 'join', input: [{ type: 'project' }, { type: 'bgp' }]}}},
+ *   { project: { preVisitor: () => ({ continue: false }), transform: projection =>
+ *     algebraFactory.createDistinct(projection) }},
+ * );
+ * const returns = { type: 'slice', input: { type: 'distinct', input: { type: 'project', input: { type: 'join', input:
+ *           [{ type: 'project' }, { type: 'bgp' }]}}}};
+ * ```
+ * @param startObject the object from which we will start the transformation,
+ *   potentially visiting and transforming its descendants along the way.
+ * @param nodeCallBacks a dictionary mapping the various operation types to objects optionally
+ *    containing preVisitor and transformer.
+ *    The preVisitor allows you to provide {@link TransformContext} for the current object,
+ *    altering how it will be transformed.
+ *    The transformer allows you to manipulate the copy of the current object,
+ *    and expects you to return the value that should take the current objects place.
+ * @return the result of transforming the requested descendant operations (based on the preVisitor)
+ * using a transformer that works its way back up from the descendant to the startObject.
+ */
 export const mapOperationStrict = transformer.transformNode.bind(transformer);
 
-export const mapOperationSub: (typeof transformer.transformNodeSpecific<'unsafe', Operation>) = <any>
-  transformer.transformNodeSpecific.bind(transformer);
+/**
+ * Transform a single operation.
+ * e.g. wrapping a distinct around the outermost project:
+ * ```ts
+ * mapOperation(
+ *   { type: 'slice', input: { type: 'project', input: { type: 'join', input: [{ type: 'project' }, { type: 'bgp' }]}}},
+ *   { project: { preVisitor: () => ({ continue: false }), transform: projection =>
+ *     algebraFactory.createDistinct(projection) }},
+ * );
+ * const returns = { type: 'slice', input: { type: 'distinct', input: { type: 'project', input: { type: 'join', input:
+ *           [{ type: 'project' }, { type: 'bgp' }]}}}};
```

Comment:
What I had in mind was something more like this for all examples:
```
mapOperation(
  {
    type: 'slice',
    input: {
      type: 'project',
      input: {
        type: 'join',
        input: [{ type: 'project' }, { type: 'bgp' }],
      },
    },
  },
  {
    [Algebra.Types.PROJECT]: {
      preVisitor: () => ({ continue: false }),
      transform: projection => algebraFactory.createDistinct(projection).
    },
  },
);
...
```
IMO, this is easier to understand when quickly glancing at the example.

Reply:
aha, okay, fully expanded. Sure